module/zfs/vdev.c

```diff
@@ -5374,6 +5402,610 @@ vdev_replace_in_progress(vdev_t *vdev)
+				ASSERT(proptype == PROP_TYPE_STRING);
+				strval = fnvpair_value_string(elem);
+				VERIFY0(zap_update(mos, objid, propname,
+				    1, strlen(strval) + 1, strval, tx));
+				spa_history_log_internal(spa, "vdev set", tx,
+				    "vdev_guid=%llu: %s=%s",
+				    (u_longlong_t)vdev_guid, nvpair_name(elem),
+				    strval);
+			} else if (nvpair_type(elem) == DATA_TYPE_UINT64) {
+				intval = fnvpair_value_uint64(elem);
+
+				if (proptype == PROP_TYPE_INDEX) {
+					const char *unused;
+					VERIFY0(vdev_prop_index_to_string(
+					    prop, intval, &unused));
+				}
+				VERIFY0(zap_update(mos, objid, propname,
+				    sizeof (uint64_t), 1, &intval, tx));
+				spa_history_log_internal(spa, "vdev set", tx,
+				    "vdev_guid=%llu: %s=%lld",
+				    (u_longlong_t)vdev_guid,
+				    nvpair_name(elem), (longlong_t)intval);
+			} else {
+				ASSERT(0); /* not allowed */
+			}
+		}
+
+	}
+
+	mutex_exit(&spa->spa_props_lock);
+}
+
+int
+vdev_prop_set(vdev_t *vd, nvlist_t *innvl, nvlist_t *outnvl)
+{
+	spa_t *spa = vd->vdev_spa;
+	nvpair_t *elem = NULL;
+	uint64_t vdev_guid;
+	nvlist_t *nvprops;
+	int error;
+
+	ASSERT(vd != NULL);
+
+	if (nvlist_lookup_uint64(innvl, ZPOOL_VDEV_PROPS_SET_VDEV,
+	    &vdev_guid) != 0)
+		return (SET_ERROR(EINVAL));
+
+	(void) nvlist_lookup_nvlist(innvl, ZPOOL_VDEV_PROPS_SET_PROPS,
+	    &nvprops);
+
+#if 0
+	if ((error = vdev_prop_validate(spa, nvprops)) != 0)
+		return (error);
+#endif
+
+	while ((elem = nvlist_next_nvpair(nvprops, elem)) != NULL) {
+		char *propname = nvpair_name(elem);
+		vdev_prop_t prop = vdev_name_to_prop(propname);
+		uint64_t intval = 0;
+		char *strval = NULL;
+
+		if (prop == VDEV_PROP_INVAL && !vdev_prop_user(propname)) {
+			error = EINVAL;
+			goto end;
+		}
+
+		if (vdev_prop_readonly(prop)) {
+			error = EROFS;
+			goto end;
+		}
+
+		/* Special Processing */
+		switch (prop) {
+		case VDEV_PROP_PATH:
+			strval = vd->vdev_path;
+			if (strval == NULL)
+				error = EROFS;
+			else if (nvpair_type(elem) != DATA_TYPE_STRING)
+				error = EINVAL;
+			if (error != 0)
+				break;
+			strval = fnvpair_value_string(elem);
+			if (strval == NULL)
+				error = EINVAL;
+			if (error != 0)
+				break;
+			spa_strfree(vd->vdev_path);
+			vd->vdev_path = spa_strdup(strval);
+			spa_config_enter(spa, SCL_CONFIG, FTAG, RW_WRITER);
+			vdev_config_dirty(vd->vdev_top);
+			spa_config_exit(spa, SCL_CONFIG, FTAG);
+			break;
+		case VDEV_PROP_ALLOCATING:
+			intval = fnvpair_value_uint64(elem);
+			if (intval != vd->vdev_noalloc)
+				break;
+			if (intval == 1)
+				error = spa_vdev_alloc(spa, vdev_guid);
+			else
+				error = spa_vdev_noalloc(spa, vdev_guid);
+			break;
+		default:
+			/* Most processing is done in vdev_sync_props */
+			break;
+		}
+end:
+		if (error != 0) {
+			intval = error;
+			vdev_prop_add_list(outnvl, propname, strval, intval, 0);
+			return (error);
+		}
+	}
+
+	return (dsl_sync_task(spa->spa_name, NULL, vdev_sync_props,
+	    innvl, 6, ZFS_SPACE_CHECK_RESERVED));
```

Comment:
I think this should be ZFS_SPACE_CHECK_EXTRA_RESERVED (at least for the allocating property) since we want the space "semantics" associated with device removal (using the allocating property) to remain consistent with the current semantics. It should be possible to set a drive as `allocating` even when we are pretty much out of space (this also could get a user out of trouble since it could "return" space to the pool).